proto-lens/src/Data/ProtoLens/Compare.hs

```diff
@@ -0,0 +1,108 @@
+-- Copyright 2016 Google Inc. All Rights Reserved.
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+-- | Matchers for comparing protocol buffer messages.
+module Data.ProtoLens.Compare (
+    -- * Matchers
+    partial,
+    ) where
+
+import Data.Constraint (Dict (..))
+import qualified Data.Map as Map
+import Lens.Family2 (Lens', over, set, view)
+import Lens.Family2.Unchecked (lens)
+import Data.ProtoLens.Message
+
+-- | Perform a partial match on a protobuffer. This corresponds to equality for
+-- primitive fields, (recursive) partial equality for required 'Message'
+-- fields, partial equality if both are set for optional fields, and partial
+-- equality of subsets for repeated and map fields.
```

Comment:
I have mixed feelings about subset equality for repeated fields and whether that satisfies the principle of least surprise.  I can recall a couple tests I've written where I did want to specify the explicit size/order of the list, I just didn't care about every field of those values.

Reply:
yeah, iirc you somehow compose matchers at google to decide which of these behaviors you want